FredrikNoren/ungit
# Switch from JSHint to ESLint

- Remove JSHint
- Add ESLint
  - Add [prettier plugin](https://github.com/prettier/eslint-plugin-prettier)
  - Add [node plugin](https://github.com/mysticatea/eslint-plugin-node)
  - Add [mocha plugin](https://github.com/lo1tuma/eslint-plugin-mocha)
- Configure linting rules specific to folder using `.eslintrc.json` files
  - `clicktest`: mocha
  - `components`: browser
  - `public/source`: browser
  - `source`: node
  - `test`: mocha

  Using files in the folders allows [editor plugins](https://github.com/microsoft/vscode-eslint) to get the correct config when a file is opened
- Use ESLint to also check formatting
  With the prettier plugin, we can check for linting issues and formatting with one command
  - Removed check-prettier-formatting grunt-task and added `npm run lint` script
  - Add lint step to CI builds explicitly
- Fix reported errors
- Use [recommended configuration](https://eslint.org/docs/user-guide/configuring#using-eslint-recommended) (of ESLint and the plugins) with the following exceptions
  - [`no-unused-vars`](https://eslint.org/docs/rules/no-unused-vars): Allow unused parameter as that seems to be current code style
  - [`no-process-exit`](https://eslint.org/docs/rules/no-process-exit): Currently used intentionally
  - [`no-control-regex`](https://eslint.org/docs/rules/no-control-regex): We have to use control characters as they appear in messages from git
  - [`no-prototype-builtins`](https://eslint.org/docs/rules/no-prototype-builtins): Only [one occurrence](https://github.com/FredrikNoren/ungit/blob/master/public/source/server.js#L59) that we can refactor later
  - [`mocha/no-mocha-arrows`](https://github.com/lo1tuma/eslint-plugin-mocha/blob/master/docs/rules/no-mocha-arrows.md): Mocha [discourages](https://mochajs.org/#arrow-functions) the use of arrow functions, which are used in the code base
  - [`mocha/no-setup-in-describe`](https://github.com/lo1tuma/eslint-plugin-mocha/blob/master/docs/rules/no-setup-in-describe.md): Can be moved to `beforeAll` blocks later
  - [`mocha/no-identical-title`](https://github.com/lo1tuma/eslint-plugin-mocha/blob/master/docs/rules/no-identical-title.md): Consider whether current naming is optimal

  The current configuration is aimed at providing the most value without requiring big refactorings for now. Future rules to consider might be:
  - [`no-var`](https://eslint.org/docs/rules/no-var): Require `const` or `let`
  - [...](https://eslint.org/docs/rules)
